pkg/controllermanager/controller/controllerdeployment/controllerdeployment.go

```diff
@@ -0,0 +1,148 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllerdeployment
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	runtimecache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet"
+)
+
+// FinalizerName is the finalizer used by this controller.
+const FinalizerName = "core.gardener.cloud/controllerdeployment"
+
+// Controller controls ManagedSeedSets.
+type Controller struct {
+	clientMap clientmap.ClientMap
+
+	reconciler reconcile.Reconciler
+
+	controllerDeploymentInformer   runtimecache.Informer
+	controllerRegistrationInformer runtimecache.Informer
+
+	controllerDeploymentQueue workqueue.RateLimitingInterface
+
+	numberOfRunningWorkers int
+	workerCh               chan int
+
+	logger *logrus.Logger
+}
+
+// New creates a new Gardener controller for ControllerDeployments.
+func New(
+	ctx context.Context,
+	clientMap clientmap.ClientMap,
+	logger *logrus.Logger,
+) (*Controller, error) {
+	gardenClient, err := clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return nil, fmt.Errorf("could not get garden client: %w", err)
+	}
+
+	controllerDeploymentInformer, err := gardenClient.Cache().GetInformer(ctx, &gardencorev1beta1.ControllerDeployment{})
+	if err != nil {
+		return nil, fmt.Errorf("could not get ControllerDeployment informer: %w", err)
+	}
+
+	controllerRegistrationInformer, err := gardenClient.Cache().GetInformer(ctx, &gardencorev1beta1.ControllerRegistration{})
+	if err != nil {
+		return nil, fmt.Errorf("could not get ControllerRegistration informer: %w", err)
+	}
+
+	reconciler := NewReconciler(clientMap, logger)
+
+	return &Controller{
+		clientMap:                      clientMap,
+		reconciler:                     reconciler,
+		controllerDeploymentInformer:   controllerDeploymentInformer,
+		controllerRegistrationInformer: controllerRegistrationInformer,
+		controllerDeploymentQueue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ManagedSeedSet"),
+		workerCh:                       make(chan int),
+		logger:                         logger,
+	}, nil
+}
+
+// Run runs the Controller until the given context is cancelled.
+func (c *Controller) Run(ctx context.Context, workers int) {
+	var waitGroup sync.WaitGroup
+
+	c.controllerDeploymentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.controllerDeploymentAdd,
+		UpdateFunc: c.controllerDeploymentUpdate,
+	})
```

Comment:
As opposed to most other controllers, why are the event handlers only added here?